network/betweenness.go

```diff
@@ -87,3 +93,57 @@ func Betweenness(g graph.Graph) map[int]float64 {
 
 	return cb
 }
+
+// BetweennessWeighted returns the betweenness centrality for nodes in the weighted
+// graph g used to construct the given shortest paths.
+//
+//  C_B(v) = \sum_{s ≠ v ≠ t ∈ V} (\sigma_{st}(v) / \sigma_{st})
+//
+func BetweennessWeighted(g graph.CostGraph, p search.ShortestPaths) map[int]float64 {
+	nodes := g.NodeList()
+	cb := make(map[int]float64, len(nodes))
```

Comment:
I've done experiments and it's the case that zero betweenness centrality states are added in Brandes' algorithm unless we add a condition with the block at https://github.com/gonum/graph/pull/64/files#diff-757c5dbafd0d4c46bdad948d072065d6R90. This is not necessary for the weighted version.

I'm beginning to think that this is probably the right thing to do - it reduces memory load, and makes it easy to do things over only between nodes or over all nodes. Here is the diff (with a few other fixes) - I won't merge it unless you agree.

```diff
diff --git a/network/betweenness.go b/network/betweenness.go
index 7933e33..d2a9b66 100644
--- a/network/betweenness.go
+++ b/network/betweenness.go
@@ -12,7 +12,7 @@ import (
        "github.com/gonum/graph/search"
 )
 
-// Betweenness returns the betweenness centrality for nodes in the unweighted graph g.
+// Betweenness returns the non-zero betweenness centrality for nodes in the unweighted graph g.
 //
 //  C_B(v) = \sum_{s ≠ v ≠ t ∈ V} (\sigma_{st}(v) / \sigma_{st})
 //
@@ -32,13 +32,10 @@ func Betweenness(g graph.Graph) map[int]float64 {
        // Also note special case for sparse networks:
        // http://wwwold.iit.cnr.it/staff/marco.pellegrini/papiri/asonam-final.pdf
 
-       nodes := g.NodeList()
-       cb := make(map[int]float64, len(nodes))
-       for _, n := range nodes {
-               cb[n.ID()] = 0
-       }
-
        var (
+               cb = make(map[int]float64)
+
+               nodes = g.NodeList()
                stack internal.NodeStack
                p     = make(map[int][]graph.Node, len(nodes))
                sigma = make(map[int]float64, len(nodes))
@@ -88,7 +85,9 @@ func Betweenness(g graph.Graph) map[int]float64 {
                                delta[v.ID()] += sigma[v.ID()] / sigma[w.ID()] * (1 + delta[w.ID()])
                        }
                        if w.ID() != s.ID() {
-                               cb[w.ID()] += delta[w.ID()]
+                               if d := delta[w.ID()]; d != 0 {
+                                       cb[w.ID()] += d
+                               }
                        }
                }
        }
@@ -96,7 +95,7 @@ func Betweenness(g graph.Graph) map[int]float64 {
        return cb
 }
 
-// BetweennessWeighted returns the betweenness centrality for nodes in the weighted
+// BetweennessWeighted returns the non-zero betweenness centrality for nodes in the weighted
 // graph g used to construct the given shortest paths.
 //
 //  C_B(v) = \sum_{s ≠ v ≠ t ∈ V} (\sigma_{st}(v) / \sigma_{st})
@@ -104,11 +103,9 @@ func Betweenness(g graph.Graph) map[int]float64 {
 // where \sigma_{st} and \sigma_{st}(v) are number of the shortest paths from s to t.
 // and the subset of those paths containing v respectively.
 func BetweennessWeighted(g graph.CostGraph, p search.ShortestPaths) map[int]float64 {
-       nodes := g.NodeList()
        cb := make(map[int]float64)
-       for _, n := range nodes {
-               cb[n.ID()] = 0
-       }
+
+       nodes := g.NodeList()
        for i, s := range nodes {
                for j, t := range nodes {
                        if i == j {
diff --git a/network/betweenness_test.go b/network/betweenness_test.go
index 177897d..bcdb008 100644
--- a/network/betweenness_test.go
+++ b/network/betweenness_test.go
@@ -37,17 +37,9 @@ var betweennessTests = []struct {
 
                wantTol: 1e-1,
                want: map[int]float64{
-                       A: 0,
                        B: 32,
-                       C: 0,
                        D: 18,
                        E: 48,
-                       F: 0,
-                       G: 0,
-                       H: 0,
-                       I: 0,
-                       J: 0,
-                       K: 0,
                },
        },
        {
@@ -78,9 +70,7 @@ var betweennessTests = []struct {
 
                wantTol: 1e-3,
                want: map[int]float64{
-                       A: 0,
                        B: 2,
-                       C: 0,
                },
        },
        {
@@ -94,11 +84,9 @@ var betweennessTests = []struct {
 
                wantTol: 1e-3,
                want: map[int]float64{
-                       A: 0,
                        B: 6,
                        C: 8,
                        D: 6,
-                       E: 0,
                },
        },
        {
@@ -112,11 +100,7 @@ var betweennessTests = []struct {
 
                wantTol: 1e-3,
                want: map[int]float64{
-                       A: 0,
-                       B: 0,
                        C: 12,
-                       D: 0,
-                       E: 0,
                },
        },
        {
@@ -129,13 +113,7 @@ var betweennessTests = []struct {
                },
 
                wantTol: 1e-3,
-               want: map[int]float64{
-                       A: 0,
-                       B: 0,
-                       C: 0,
-                       D: 0,
-                       E: 0,
-               },
+               want:    map[int]float64{},
        },
 }
 
@@ -156,7 +134,12 @@ func TestBetweenness(t *testing.T) {
                got := Betweenness(g)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
-                       if !floats.EqualWithinAbsOrRel(got[n], test.want[n], test.wantTol, test.wantTol) 
+                       wantN, gotOK := got[n]
+                       gotN, wantOK := test.want[n]
+                       if gotOK != wantOK {
+                               t.Errorf("unexpected betweenness result for test %d, node %d", i, n)
+                       }
+                       if !floats.EqualWithinAbsOrRel(gotN, wantN, test.wantTol, test.wantTol) {
                                t.Errorf("unexpected betweenness result for test %d:\ngot: %v\nwant:%v",
                                        i, orderedFloats(got, prec), orderedFloats(test.want, prec))
                                break
@@ -187,12 +170,16 @@ func TestBetweennessWeighted(t *testing.T) {
                }
 
                got := BetweennessWeighted(g, p)
-               want := Betweenness(g)
                prec := 1 - int(math.Log10(test.wantTol))
                for n := range test.g {
-                       if !floats.EqualWithinAbsOrRel(got[n], want[n], test.wantTol, test.wantTol) {
+                       gotN, gotOK := got[n]
+                       wantN, wantOK := test.want[n]
+                       if gotOK != wantOK {
+                               t.Errorf("unexpected betweenness existence for test %d, node %d", i, n)
+                       }
+                       if !floats.EqualWithinAbsOrRel(gotN, wantN, test.wantTol, test.wantTol) {
                                t.Errorf("unexpected betweenness result for test %d:\ngot: %v\nwant:%v",
-                                       i, orderedFloats(got, prec), orderedFloats(want, prec))
+                                       i, orderedFloats(got, prec), orderedFloats(test.want, prec))
                                break
                        }
                }
```